### ISSUE TITLE
Use SIGTERM instead of SIGKILL when ending processes, allow attachSimplePort=0

### ISSUE THREAD

**Issue by alexkli** [microsoft/vscode-js-debug#630]
**Describe the bug**

(Note sure if bug or feature request is most appropriate, this might be a somewhat special case!)

[wskdebug](https://github.com/apache/openwhisk-wskdebug) is a debugging tool for Apache Openwhisk (a serverless platform), which starts a container with a nodejs process inside that will have the actual debug port to which the VS Code debugger should connect. I am using this in VS Code as `runtimeExecutable` and have been intercepting the `--inspect-brk=NNNN` argument that was set by the previous debugger to use this port on the container that gets started. VS Code then did connect to the container, which was great.

However, it seems the new debugger uses a different mechanism. Apparently it sets `NODE_OPTIONS` with a `--require` that loads a JS file into the node process that apparently sets up the debugging from "inside" the process.

**The problem is that now it debugs the `wskdebug` "intermediary" and not the target container.**

Is there a way in the new debugger I can disable that and just get hold off the port number (which VS code randomly chooses, i.e. in case there are multiple debug sessions going on)?

See also https://github.com/apache/openwhisk-wskdebug/issues/74

On a related note, the new debugger seems to kill the process at the end immediately (SIGKILL style). The previous debugger sent a `SIGTERM` so that the process had a chance to cleanup and properly exit. In `wskdebug` case this is used to stop the docker container amongst other things. With the new debugger this is no longer possible and thus the container with the open debug port is left behind, so that the next run usually fails with "debug port already in use". This is also an artifact of the above issue since always the same default debug port (9229) is picked now since `--inspect-brk=NNNN` with a unique port NNNN is missing.

**To Reproduce**
1. Run launch config with `wskdebug` as [described here](https://github.com/apache/openwhisk-wskdebug#nodejs-visual-studio-code)
2. Notice debugger connects to `wskdebug` itself, but not the container it starts
3. Cannot hit any breakpoints since it's debugging the wrong process

**Log File**

[vscode-debugadapter-9.json.gz](https://github.com/microsoft/vscode-js-debug/files/4930390/vscode-debugadapter-9.json.gz)

**VS Code Version:** 1.47.1

**Comment by connor4312** [comment]
Thanks for the details. I think your scenario and fix is covered in https://github.com/microsoft/vscode-js-debug/issues/586#issuecomment-658958272. However I'll leave this open to use SIGTERM instead of KILL.

**Comment by alexkli** [comment]
@connor4312 Thanks so much for your quick reply!

I tried the nightly + `attachSimplePort` but it did not fix the issue. VS code still invokes the program _without_ the `--inspect-brk=NNNN` argument.

How exactly do I set it in the `launch.json`? I have this but the editor underlines it with a warning `Incorrect type. Expected "number"`:

```
      "port": "attachSimplePort"
```

I am using version `v2020.7.1517` of `@id:ms-vscode.js-debug-nightly`.

**Comment by connor4312** [comment]
The nightly where `--inspect-brk` will not be filtered is today's nightly, which will be available in about two hours from now 🙂 

The change you want in your package.json is 

```diff
{
  "type": "pwa-node", // if only "node", VS Code will complain
  "request": "launch",
  "name": "My launch",
+ "attachSimplePort": 9229,
```

**Comment by alexkli** [comment]
Ah, sorry, I missed the "today's nightly" part!

Regarding:
```
"attachSimplePort": 9229
```

Unfortunately that forces me to pick a port and hardcode it into each launch config, which puts some burden on the user.

Especially if there are multiple launch configs (which is likely in our use case, one might debug multiple serverless functions in parallel), which all would need a different port number. And also if another debug session is going on or simply a leftover nodejs/container is still running in the background, the auto selection nicely handled those issues.

**Comment by connor4312** [comment]
Added in the linked commit. Just assign attachSimplePort to 0, and js-debug will automatically pick a port and add --inspect-brk as it did previously. Note that with this you *must* now have the type be `pwa-node`, otherwise it won't go through js-debug's config resolver where this logic happens.

**Comment by alexkli** [comment]
Can confirm that attachSimplePort works in the nightly 🥳. 

For reference, here is the `launch.json` snippet I used:
```json
      "type": "pwa-node",
      "request": "launch",
      "name": "wskdebug",
      "attachSimplePort": 0,
```

**Comment by alexkli** [comment]
@connor4312 Hi there, I was wondering if the changes for `pwa-node` and `attachSimplePort ` are available in the July 2020 VS Code release aka the 1.48.x version?

**Comment by connor4312** [comment]
Yep, they are

**Comment by jkoput** [comment]
Is the change back to SIGTERM/SIGINT instead of SIGKILL on the roadmap?  This is impacting a current development; would we be better off looking at a different IDE?

**Comment by connor4312** [comment]
I do not have an ETA for this. However PRs are welcome.

**Comment by zgz682000** [comment]
Why deprecate sending SIGTERM/SIGINT, without them I don't know how to gracefully shutdown my node server when debuging with vscode. Have anyone ideas?

**Comment by zgz682000** [comment]
And with the config `"attachSimplePort":0,`, my forked process will report inspect port error
`Starting inspector on 127.0.0.1:21154 failed: address already in use`

**Comment by connor4312** [comment]
You can now set `"killBehavior": "polite"` to use SIGTERM on posix. On windows this has the effect of omitted the `/F` force flag from taskkill.exe.

**Comment by roblourens** [comment]
Is this like the thing that's supposed to happen with the terminateRequest? From what I remember, the DA can opt in to first get a terminateRequest whenthe user clicks the stop button, then if the program doesn't terminate, the user can click the stop button a second time and the DA can try again with the stronger signal. https://github.com/Microsoft/vscode/issues/57018#issuecomment-415156959

**Comment by roblourens** [comment]
The current behavior just leaves the process running (if it intercepts the signal), is that expected?

**Comment by connor4312** [comment]
> The current behavior with killBehavior=polite leaks the process (if it intercepts the signal), that seems not great

I think that's up to the user to deal with in that case. If they opt in so that js-debug terminates their process gracefully and it fails to shut down as a result of SIGTERM, that's an issue that needs to be fixed in their program. We could introduce a heuristic that forcefully kills the process if it remains alive after some timeout, and I think that would be good if SIGTERM was the default behavior, but because it's opt-in I'm not sure that level of extra configuration for the timeout is needed.

**Comment by roblourens** [comment]
That's fine. How about https://github.com/microsoft/vscode-js-debug/issues/630#issuecomment-738988895 though?

**Comment by connor4312** [comment]
It looks like can happen, but js-debug does not have `supportsTermianteRequest`. I'll open a new issue for that since this thread has several different branches, but I think that's better than specifying the kill behavior.
